components/brave_wallet/browser/brave_wallet_pin_service.cc

```diff
@@ -0,0 +1,671 @@
+  switch (status) {
+    case mojom::TokenPinStatusCode::STATUS_NOT_PINNED:
+      return "not_pinned";
+    case mojom::TokenPinStatusCode::STATUS_PINNED:
+      return "pinned";
+    case mojom::TokenPinStatusCode::STATUS_PINNING_IN_PROGRESS:
+      return "pinning_in_progress";
+    case mojom::TokenPinStatusCode::STATUS_UNPINNING_IN_PROGRESS:
+      return "unpining_in_progress";
+    case mojom::TokenPinStatusCode::STATUS_UNPINNING_FAILED:
+      return "unpining_failed";
+    case mojom::TokenPinStatusCode::STATUS_PINNING_FAILED:
+      return "pining_failed";
+    case mojom::TokenPinStatusCode::STATUS_PINNING_PENDING:
+      return "pinning_pendig";
+    case mojom::TokenPinStatusCode::STATUS_UNPINNING_PENDING:
+      return "unpinning_pendig";
+  }
+  NOTREACHED();
+  return "";
+}
+
+std::string ErrorCodeToString(
+    const mojom::WalletPinServiceErrorCode& error_code) {
+  switch (error_code) {
+    case mojom::WalletPinServiceErrorCode::ERR_WRONG_TOKEN:
+      return "ERR_WRONG_TOKEN";
+    case mojom::WalletPinServiceErrorCode::ERR_NON_IPFS_TOKEN_URL:
+      return "ERR_NON_IPFS_TOKEN_URL";
+    case mojom::WalletPinServiceErrorCode::ERR_FETCH_METADATA_FAILED:
+      return "ERR_FETCH_METADATA_FAILED";
+    case mojom::WalletPinServiceErrorCode::ERR_WRONG_METADATA_FORMAT:
+      return "ERR_WRONG_METADATA_FORMAT";
+    case mojom::WalletPinServiceErrorCode::ERR_ALREADY_PINNED:
+      return "ERR_ALREADY_PINNED";
+    case mojom::WalletPinServiceErrorCode::ERR_NOT_PINNED:
+      return "ERR_NOT_PINNED";
+    case mojom::WalletPinServiceErrorCode::ERR_PINNING_FAILED:
+      return "ERR_PINNING_FAILED";
+  }
+  NOTREACHED();
+  return "";
+}
+
+BraveWalletPinService::BraveWalletPinService(
+    PrefService* prefs,
+    JsonRpcService* service,
+    ipfs::IpfsLocalPinService* local_pin_service)
+    : prefs_(prefs),
+      json_rpc_service_(service),
+      local_pin_service_(local_pin_service) {}
+
+BraveWalletPinService::~BraveWalletPinService() {}
+
+mojo::PendingRemote<mojom::WalletPinService>
+BraveWalletPinService::MakeRemote() {
+  mojo::PendingRemote<WalletPinService> remote;
+  receivers_.Add(this, remote.InitWithNewPipeAndPassReceiver());
+  return remote;
+}
+
+void BraveWalletPinService::Bind(
+    mojo::PendingReceiver<mojom::WalletPinService> receiver) {
+  receivers_.Add(this, std::move(receiver));
+}
+
+void BraveWalletPinService::AddObserver(
+    ::mojo::PendingRemote<mojom::BraveWalletPinServiceObserver> observer) {
+  observers_.Add(std::move(observer));
+}
+
+// static
+std::string BraveWalletPinService::GetPath(
+    const absl::optional<std::string>& service,
+    const BlockchainTokenPtr& token) {
+  return base::StrCat({kNftPart, ".", service.value_or(kLocalService), ".",
+                       base::NumberToString(static_cast<int>(token->coin)), ".",
+                       token->chain_id, ".", token->contract_address, ".",
+                       token->token_id});
```

Comment:
(same for BraveWalletPinService::GetTokens)
Is it safe to make pref path this way? What if any part of it has a dot?
Do we really need keep such hierarchy in prefs? Maybe just an array of {service, coin, chain_id, etc...} ?

Reply:
It is needed some times to resolve tokens by group like service where they are pinned.
I've added protection in GetPath method